File: app/src/main/java/com/github/teranes10/androidutils/utils/SignalRUtil.java
package com.github.teranes10.androidutils.utils;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.os.SystemClock;
import android.util.Log;

import com.microsoft.signalr.HttpHubConnectionBuilder;
import com.microsoft.signalr.HubConnection;
import com.microsoft.signalr.HubConnectionBuilder;
import com.microsoft.signalr.HubConnectionState;

import java.util.concurrent.CompletableFuture;

import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.disposables.Disposable;

public abstract class SignalRUtil {
    private static final String TAG = "SignalR";
    private static final int SERVER_TIMEOUT = 30 * 1000;
    private static final int KEEP_ALIVE_INTERVAL = 20 * 1000;
    private static final int[] CONNECTION_RETRIES = new int[]{0, 10000, 20000, 30000, 60000};
    private static final int RESET_CONNECTION_RETRY_IN = 5 * 60 * 1000;

    private final Context context;
    private final SignalRStatusListener listener;
    private HubConnection hubConnection;
    private long lastReconnectedAt = 0;
    private int currentConnectionTry = 0;
    private boolean isConnecting;
    private CompletableFuture<HubConnection> connectionFuture;
    private Disposable connectionDisposable;

    public enum ConnectionStatus {
        Disconnected,
        Connecting,
        Connected,
        Reconnecting,
        NotConnected
    }

    private final ConnectionUtil.ConnectionListener internetConnectionListener = isConnected -> {
        if (isConnected) {
            updateStatus(ConnectionStatus.NotConnected);
            if (!isConnecting) {
                connect();
            }
        }
    };

    public interface SignalRStatusListener {
        void onSignalRConnectionStatusChanged(ConnectionStatus status);
    }

    public SignalRUtil(Context context, SignalRStatusListener listener) {
        this.context = context;
        this.listener = listener;
    }

    private void updateStatus(ConnectionStatus status) {
        if (listener != null) {
            listener.onSignalRConnectionStatusChanged(status);
        }
    }

    private HubConnection createConnection() {
        if (Utils.isNullOrEmpty(setUrl())) {
            return null;
        }

        HttpHubConnectionBuilder builder = HubConnectionBuilder.create(setUrl());
        if (!Utils.isNullOrEmpty(setToken())) {
            builder.withAccessTokenProvider(Single.defer(() -> Single.just(setToken())));
        }

        HubConnection connection = builder.build();
        connection.setServerTimeout(SERVER_TIMEOUT);
        connection.setKeepAliveInterval(KEEP_ALIVE_INTERVAL);

        setListeners(connection);

        connection.onClosed(e -> {
            Log.e(TAG, "onClosed: " + setTag() + ": " + e.getLocalizedMessage());
            onConnectionClosed();
            updateStatus(ConnectionStatus.Reconnecting);
            reconnect();
        });

        return connection;
    }

    private void reconnect() {
        try {
            long elapsed = SystemClock.elapsedRealtime() - lastReconnectedAt;
            if (elapsed > RESET_CONNECTION_RETRY_IN) {
                currentConnectionTry = 0;
            }

            if (currentConnectionTry < CONNECTION_RETRIES.length) {
                lastReconnectedAt = SystemClock.elapsedRealtime();
                new Handler(Looper.getMainLooper()).postDelayed(() -> {
                    if (!isConnecting) {
                        connect();
                    }
                }, CONNECTION_RETRIES[currentConnectionTry]);

                currentConnectionTry++;
            } else {
                onConnectionClosed();
                updateStatus(ConnectionStatus.Disconnected);
            }
        } catch (Exception e) {
            Log.e(TAG, "reconnect: " + e.getLocalizedMessage());
        }
    }

    private void startInternetConnectionObserver() {
        DefaultConnectionUtil instance = DefaultConnectionUtil.getInstance(context);
        if (instance != null) {
            instance.addListener(internetConnectionListener);
        }
    }

    private void stopInternetConnectionObserver() {
        DefaultConnectionUtil instance = DefaultConnectionUtil.getInstance(context);
        if (instance != null) {
            instance.removeListener(internetConnectionListener);
        }
    }

    public String setTag() {
        return getClass().getSimpleName();
    }

    public String setToken() {
        return null;
    }

    protected abstract String setUrl();

    protected abstract void setListeners(HubConnection connection);

    public CompletableFuture<HubConnection> connect() {
        if (isConnected(hubConnection)) {
            return CompletableFuture.completedFuture(hubConnection);
        }

        if (isConnecting && connectionFuture != null) {
            return connectionFuture.thenApply(hubConnection -> {
                if (isConnected(hubConnection)) {
                    Log.i(TAG, "ConnectionId: " + hubConnection.getConnectionId());
                    return hubConnection;
                }
                return null;
            }).exceptionally(e -> {
                Log.e(TAG, "Error while waiting for connection: " + setTag(), e);
                return null;
            });
        }

        if (hubConnection == null) {
            hubConnection = createConnection();
        }

        updateStatus(ConnectionStatus.Connecting);
        isConnecting = true;
        connectionFuture = new CompletableFuture<>();

        dispose();
        connectionDisposable = hubConnection.start().subscribe(
                () -> {
                    connectionFuture.complete(hubConnection);
                    onConnected();
                },
                e -> {
                    Log.e(TAG, "Connection error: " + e.getLocalizedMessage());
                    connectionFuture.complete(null);
                    onConnectionClosed();
                });

        return connectionFuture;
    }

    protected void onConnected() {
        isConnecting = false;
        connectionFuture = null;

        if (isConnected()) {
            Log.i(TAG, "Connected.");
            onConnected();
            updateStatus(ConnectionStatus.Connected);
            stopInternetConnectionObserver();
        } else {
            onConnectionClosed();
        }
    }

    protected void onConnectionClosed() {
        isConnecting = false;
        connectionFuture = null;

        updateStatus(ConnectionStatus.Disconnected);
        startInternetConnectionObserver();
    }

    public boolean isConnecting() {
        return isConnecting;
    }

    public boolean isConnected(HubConnection connection) {
        return connection != null && connection.getConnectionState() == HubConnectionState.CONNECTED;
    }

    public boolean isConnected() {
        return isConnected(hubConnection);
    }

    public boolean canConnect() {
        return !isConnecting && !isConnected();
    }

    public CompletableFuture<HubConnection> getConnection(String tag) {
        if (isConnecting && connectionFuture != null) {
            return connectionFuture.thenApply(hubConnection -> {
                if (isConnected(hubConnection)) {
                    Log.i(TAG, "ConnectionId: " + hubConnection.getConnectionId());
                    return hubConnection;
                }
                return null;
            }).exceptionally(e -> {
                Log.e(TAG, "Error waiting for connection: " + tag, e);
                return null;
            });
        }

        return CompletableFuture.completedFuture(isConnected() ? hubConnection : null);
    }

    private Disposable connectionCloseDisposable;

    public CompletableFuture<Void> close() {
        Log.i(TAG, "Closing connection.");
        CompletableFuture<Void> future = new CompletableFuture<>();

        if (isConnected()) {
            dispose();
            connectionCloseDisposable = hubConnection.stop().subscribe(() -> {
                future.complete(null);
                Log.i(TAG, "Connection closed.");
            }, future::completeExceptionally);
        }

        return future;
    }

    public void dispose() {
        if (connectionDisposable != null && !connectionDisposable.isDisposed()) {
            connectionDisposable.dispose();
        }
        if (connectionCloseDisposable != null && !connectionCloseDisposable.isDisposed()) {
            connectionCloseDisposable.dispose();
        }
    }
}

File: app/src/main/java/com/github/teranes10/androidutils/utils/SignalRUtil.kt
package com.github.teranes10.androidutils.utils

import android.os.SystemClock
import android.util.Log
import com.github.teranes10.androidutils.extensions.ExceptionExtensions.displayMessage
import com.github.teranes10.androidutils.extensions.RxJavaExtensions.await
import com.github.teranes10.androidutils.extensions.SignalRExtensions.invokeSuspend
import com.microsoft.signalr.HubConnection
import com.microsoft.signalr.HubConnectionBuilder
import com.microsoft.signalr.HubConnectionState
import io.reactivex.rxjava3.core.Single
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

abstract class SignalRUtil(protected val scope: CoroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())) {

    companion object {
        private const val TAG = "SignalR"
        private const val SERVER_TIMEOUT = 30000L
        private const val KEEP_ALIVE_INTERVAL = 20000L
        private val CONNECTION_RETRIES = listOf(0L, 10000L, 20000L, 30000L, 60000L, 120000L)
        private const val RESET_CONNECTION_RETRY_IN = 5 * 60 * 1000L
    }

    open class ConnectionStatus(val message: String) {
        data object Disconnected : ConnectionStatus("Not connected")
        data object Connecting : ConnectionStatus("Attempting to connect")
        data object Connected : ConnectionStatus("Connected successfully")
        data object Reconnecting : ConnectionStatus("Reconnecting...")
    }

    protected val _events = MutableStateFlow<ConnectionStatus>(ConnectionStatus.Disconnected)
    val events = _events.asStateFlow()

    private var hubConnection: HubConnection? = null
    private var lastReconnectedAt = 0L
    private var currentConnectionTry = 0
    private val isConnecting = AtomicBoolean(false)
    private val connectionFuture: AtomicReference<Deferred<Unit>?> = AtomicReference(null)
    private val connectionMutex = Mutex()

    val isConnected: Boolean get() = isConnected(hubConnection)

    private fun isConnected(connection: HubConnection?): Boolean {
        return connection?.connectionState == HubConnectionState.CONNECTED
    }

    private fun updateStatus(status: ConnectionStatus) {
        _events.update { status }
    }

    private fun createConnection(): HubConnection? {
        val url = setUrl()
        if (url.isNullOrEmpty()) return null

        val builder = HubConnectionBuilder.create(url)
        setToken()?.let { token -> builder.withAccessTokenProvider(Single.defer { Single.just(token) }) }

        return builder.build().apply {
            serverTimeout = SERVER_TIMEOUT
            keepAliveInterval = KEEP_ALIVE_INTERVAL
            setListeners(this)
            onClosed {
                Log.e(TAG, "onClosed (start reconnection): ${setTag()}: ${it.displayMessage}")
                scope.launch { reconnect() }
            }
        }
    }

    private suspend fun reconnect() {
        val elapsed = SystemClock.elapsedRealtime() - lastReconnectedAt
        if (elapsed > RESET_CONNECTION_RETRY_IN) {
            currentConnectionTry = 0
        }

        if (currentConnectionTry < CONNECTION_RETRIES.size) {
            lastReconnectedAt = SystemClock.elapsedRealtime()

            if (!isConnecting.get()) {
                delay(CONNECTION_RETRIES[currentConnectionTry])
                connect(true, "onReconnect")
            }

            currentConnectionTry++
        } else {
            onConnectionClosed()
        }
    }

    protected abstract fun setUrl(): String?
    protected abstract fun setListeners(connection: HubConnection)
    protected open fun setTag(): String = javaClass.simpleName
    protected open fun setToken(): String? = null

    open fun onConnected() {
        Log.i(TAG, "Connected.")
        updateStatus(ConnectionStatus.Connected)
    }

    open fun onConnectionClosed() {
        Log.e(TAG, "onConnectionClosed: ")
        updateStatus(ConnectionStatus.Disconnected)
    }

    suspend fun connect(reconnect: Boolean = false, tag: String = "") {
        Log.i(TAG, "connect: $tag")

        if (isConnected) {
            Log.i(TAG, "connect: already connected")
            return
        }

        if (!connectionMutex.tryLock(SERVER_TIMEOUT)) {
            Log.e(TAG, "Unable to acquire lock to connect")
            return
        }

        try {
            if (isConnected) {
                Log.i(TAG, "connect: already connected (inside critical section)")
                return
            }

            Log.i(TAG, "connect: connecting")
            isConnecting.set(true)
            updateStatus(if (reconnect) ConnectionStatus.Reconnecting else ConnectionStatus.Connecting)

            if (hubConnection == null) {
                hubConnection = createConnection()
            }

            try {
                hubConnection?.start()?.await()
                onConnected()
            } catch (e: Exception) {
                Log.e(TAG, "Connection error: ${e.displayMessage}", e)
                onConnectionClosed()
            } finally {
                isConnecting.set(false)
            }
        } finally {
            connectionMutex.unlock()
        }
    }

    suspend fun getConnection(): HubConnection? {
        if (isConnected) {
            return hubConnection
        }

        if (isConnecting.get()) {
            connectionMutex.withLock {
                return if (isConnected) hubConnection else null
            }
        }

        return null
    }

    suspend fun <T : Any> invoke(
        returnType: Class<T>,
        method: String,
        vararg args: Any
    ): T? = withContext(Dispatchers.IO) {
        try {
            val connection = getConnection()
            if (connection == null) {
                Log.e(TAG, "$method: No connection.")
                return@withContext null
            }

            Log.i(TAG, "$method: ${args.joinToString(", ")}")
            val res = connection.invokeSuspend(returnType, method, *args)
            Log.i(TAG, "$method: ${args.joinToString(", ")} :: $res")
            return@withContext res
        } catch (e: Exception) {
            Log.e(TAG, "$method: ", e)
            return@withContext null
        }
    }

    suspend fun close() = withContext(Dispatchers.IO) {
        if (isConnected) {
            try {
                hubConnection?.stop()?.blockingAwait()
                Log.i(TAG, "Connection closed.")
            } catch (e: Exception) {
                Log.e(TAG, "Error closing connection: ${e.displayMessage}", e)
            }
        }
    }
}